### ISSUE TITLE
Define Branching Plan for aspnet/JitBench

### ISSUE THREAD

**Issue by rynowak** [aspnet/JitBench#25]
A recent PR generated a lot of discussion about what the scenarios and requirements are for startup testing. I want to put forth a plan here and make sure that we've covered everyone's concerns.

## Music Store App Requirements

I want to first point out a few requirements *just from the standpoint of the app* that help drive this discussion in a more practical version. 

First, the Music Store app has a strong degree of coupling to a release of ASP.NET. We break things more frequently than the BCL, and ASP.NET 2.0 (current development) has already numerous breaking changes that impact the Music Store app. So, because the source code of the app is fixed in a given branch, it means that each branch needs to target *first and foremost* a specific ASP.NET release (or current development). 

Secondly, the app itself has very little coupling to the release of the .NET shared framework besides what ASP.NET requires. We (ASP.NET) are generally going to be sticking with the `netstandard1.6` target for most of our libraries. Where we do get into trouble here with coupling to .NET is the SDK. Any changes which come down the pipe that break the project file require new SDKs. Fortunately `clrjit.dll` lives in the shared framework. I believe `dotnet cache` also uses the `clrjit.dll` from the shared framework - but we might want to confirm that. 

**TLDR** - Music Store is is coupled to a release of ASP.NET but is very flexible with respect to the version of the shared-framework/JIT.

## Requirements from the Runtime Team

Here's a summary of the requirements I've been hearing from various runtime team members. If you don't see your feedback captured here, let me know!

- We want to target a stable version of ASP.NET so we can easier understand the impact of runtime changes. We also want to target latest development of ASP.NET, but this is lower priority.
- We want to make sure that R2R/crossgen and the app are run with the same version of the JIT. Running a new JIT for runtime JIT but an older JIT for R2R/crossgen only give us half the picture.
- We want to be able to select the version of `dotnet`/shared-framework for a given run.
- We want to be able to run on all platforms/architectures where `dotnet` runs

## The proposal

My proposal is that we have a branch-per-ASP.NET release as well as a special branch for the current development of ASP.NET. The code and project file of MusicStore will be based on the specific ASP.NET release targeted by its resident branch. Each branch will support the minimum version of the `dotnet` shared framework supported by that ASP.NET release and all newer versions of `dotnet` - including latest development. This supports the requirement to test changes in the runtime which keeping the ASP.NET libraries and app fixed.

**Example**
Once `2.0.0-preview1` is released, the `release-2.0.0-preview1` branch will support the `2.0.0-preview1` release of ASP.NET and the `2.0.0-preview1` release of the SDK/shared-framework.  It will also be possible to use any **newer** version of the SDK/shared-framework using this branch. 

The 2.0.0 release is a major change from the 1.0.0 dotnet/ASP.NET releases. Much of the functionality infrastructure we use (such as dotnet cache) is not available in in the older releases. For that reason I propose that we create/support branches of the 1.0.0 family of releases only as needed by our business requirements. Assume that everything mentioned here applies to the 2.0.0 family of releases and all subsequent releases. When we maintain a branch of the 1.0.0 family it will be somewhat ad-hoc and will not support all of the knobs and dials of the newer bits.

In each branch starting with `release-2.0.0-preview1` we'll support performing the crossgen/R2R operation in the test environment. This satisfies the requirement that we use a single runtime version for both runtime JIT and crossgen R2R.

In each branch starting with `release-2.0.0-preview1` we'll support performing the full set of options for testing in all architectures and platforms supported by dotnet. This support will not require separate branches, just separate scripts where needed (powershell vs bash)

In each branch starting with `release-2.0.0-preview1` we'll support a number of variations on the publishing and running of MusicStore. None of these changes require a significantly different workflow, they just require testing and documenting them. This can include but is not limited to:
- Multi-tiered jitting
- View Compilation on publish (ASP.NET production scenario)
- Running without R2R/crossgen binaries

## Unknowns

It's not clear yet what our (ASP.NET) requirements will be for what we want to automate and measure. I would suppose that we would be much more interested in the end-to-end scenarios - *what's the difference between 1.0.0 of dotnet/ASP.NET and 2.0.0 of dotnet/ASP.NET* - *what's the impact of precompiled views?* - *what's the impact of tiered JIT?*. All of these things are just twists on how to publish/run the MusicStore app, and so I don't think they need to affect our overall strategy.

It's not yet clear to me to what degree runtime team members what to *hack* a local build of the shared framework and iterate locally. The concern here is that `dotnet cache` goes out and restores a build of the shared framework to use for running crossgen. If we need to be able to do this locally as part of a dev inner loop, then we should request this feature from the SDK team as an additional parameter.

## Appendix A: Original Feedback

I think I've addressed all of these points in the strategy and discussion. If you think your concerns are not sufficiently addressed then say so and we'll figure it out.

From @AndyAyersMS 

>Aspnetcore-dev churns pretty frequently, with only a day or two of packages kept around. The corequisite churns CLI somewhat less frequently, not sure what the policy is there for updates.

>So with fetch-latest policy we run the risk of being unable to differentiate perf changes caused by asp changes from those caused by jit/runtime/framework changes.

>We also may be unable to root-cause and repro past runs because older packages aren't available.

>For the jit team's use it would be more ideal if the aspnet/cli side of things did not update automatically but could remain stable for some predictable interval -- a month would be good. We can have a separate "scout ahead" run that uses latest bits if we want to try and track both.

>I also agree with Joseph that running with some or perhaps even most of the code produced with an older jit is going to cause further confusion. We'd need to be able to regenerate all these bits with a newer jit / runtime / framework. Otherwise we might have latent perf surprises when the pregenerated code picks up a newer jit.

---------------------

From @JosephTremoulet 

> @AndyAyersMS I think of the 'dev' branch of this repo (which is what this PR is targeting) as the "scout ahead" run -- the simplest place to get a latestAspnet-with-latestCoreclr run to do an A/B comparison for a particular change.

> I agree that doesn't make it ideal for analyzing the impact over time of jit changes specifically. Presuming we want both, I don't know if we should just use the master branch (which is tied to released aspnet) for that or have CI runs of the dev branch against fixed aspnet snapshots like you're describing.

> Having the flexibility in the dev branch to specify a version (CI could use this) or just grab latest (simplifies joint aspnet/jit development) would probably be ideal. I'd argue the default should be to grab latest.

-----------------

From @lt72 

> Andy has a good point. We should definitely have a run configuration with a stable framework and latest JIT. The main reason we created this repo is to test the startup (most important) and steady state perfs of the latest ASP.NET Core w and w/o tiered jitting. Startup is a target for v2, so it should definitely take precedence. Considering that we did not manage to have a stable run for a while, I would suggest to make this work predictably in one configuration and work on other configurations immediately after.

> RE: platform priorities: from highest to lowest:

>Win x64
>Linux x64
> Win x86
> Linux x86
> @rynowak and @davmason : what it would take to create another run configuration that mantains a stable framework as Andy suggested? do we need another branch?

**Comment by rynowak** [comment]
For the immediate action items here, I suggest that we keep making progress on `dotnet cache` and xplat scripting on the `dev` branch.

The `2.0.0-preview1` is only a few weeks away, so when that is available we branch that off `dev` and just hardcode a few version numbers. Then this branch will become a more stable target for automation. 

Once we have this accomplished we should gather more input and make a decision about the 1.0.0 branches.

**Comment by JosephTremoulet** [comment]
SGTM

**Comment by rynowak** [comment]
This is done now. Use the `rel/2.0.0-preview1` branch for a stable version of the ASP.NET libraries
